test_ria_basics.py::test_binary_data: wants interaction

Prior issue (closed as addressed) for this test was #4371 , suggests something about Authorization (didn't check in detail) hence might indeed relate to the desire of interaction.

Happened in the course of testing https://github.com/datalad/datalad/pull/7126 which AFAIK does not relate, happened on travis in a single run of the heavy debug logging etc:

```
=========================== short test summary info ============================
FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'download_url',
  'error_message': "'ConsoleLog' object has no attribute 'yesno'",
  'exception': 'ConsoleLog' object has no attribute 'yesno' [download_url.py:__call__:202,base.py:download:527,base.py:access:219,base.py:_handle_authentication:286,base.py:_enter_credentials:319,__init__.py:__getattribute__:101,base.py:access:174,base.py:_download:427,http.py:get_downloader_session:622,http.py:check_response_status:124],
  'exception_traceback': '[download_url.py:__call__:202,base.py:download:527,base.py:access:219,base.py:_handle_authentication:286,base.py:_enter_credentials:319,__init__.py:__getattribute__:101,base.py:access:174,base.py:_download:427,http.py:get_downloader_session:622,http.py:check_response_status:124]',
  'message': "AttributeError('ConsoleLog' object has no attribute 'yesno')",
  'path': '/tmp/datalad_temp__test_binary_datathev6l40/dicomfile',
  'status': 'error',
  'type': 'file'}]
= 1 failed, 1162 passed, 46 skipped, 3 xfailed, 17 warnings in 1023.67s (0:17:03) =
The command "http_proxy= PATH=$PWD/../tools/coverage-bin:$PATH $PYTEST_WRAPPER python -m pytest "${PYTEST_OPTS[@]}" -c ../tox.ini -n 2 -m "${PYTEST_SELECTION:+$PYTEST_SELECTION_OP($PYTEST_SELECTION) and }not(turtle)" --doctest-modules --cov=datalad --cov-report= --pyargs $TESTS_TO_PERFORM" exited with 1.
0.00s$ cd ..
```

Seems to be relatively frequent encounter, and only on travis (looked only at this month):

```
(git)smaug:/mnt/datasets/datalad/ci/logs/2022/10[master]git
$> git grep '^FAILED.*test_binary_data'
11/pr/7068/08876f3/travis-14700-failed/9.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
12/pr/7002/16a162d/travis-14714-failed/8.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
19/pr/7094/a88c6e1/travis-14761-failed/2.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
19/push/master/1de6d08/travis-14764-failed/2.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
21/push/maint/e94a49e/travis-14772-failed/4.txt:FAILED ../distributed/tests/test_ria_basics.py::test_binary_data - datalad.su...
24/push/maint/969bb82/travis-14794-failed/9.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
git grep '^FAILED.*test_binary_data'  4.04s user 3.52s system 1016% cpu 0.743 total
```
and seems to happen not only in the "heavy debug" run

--- a/changelog.d/pr-7142.md
+++ b/changelog.d/pr-7142.md
@@ -1,0 +1,7 @@
+### 🐛 Bug Fixes
+
+- HTTP download treated '429 - too many requests' as an authentication issue and
+  was consequently trying to obtain credentials.
+  Fixes [#7129](https://github.com/datalad/datalad/issues/7129) via
+  [PR #YYY](https://github.com/datalad/datalad/pull/YYY)
+  (by [@bpoldrack](https://github.com/bpoldrack))

--- a/changelog.d/pr-7142.md
+++ b/changelog.d/pr-7142.md
@@ -3,5 +3,5 @@
 - HTTP download treated '429 - too many requests' as an authentication issue and
   was consequently trying to obtain credentials.
   Fixes [#7129](https://github.com/datalad/datalad/issues/7129) via
-  [PR #YYY](https://github.com/datalad/datalad/pull/YYY)
+  [PR #7129](https://github.com/datalad/datalad/pull/7129)
   (by [@bpoldrack](https://github.com/bpoldrack))

--- a/datalad/downloaders/base.py
+++ b/datalad/downloaders/base.py
@@ -179,7 +179,7 @@ class BaseDownloader(object, metaclass=ABCMeta):
                 if hasattr(e, 'status') and e.status == 429:
                     # Too many requests.
                     # We can retry by continuing the loop.
-                    time.sleep(0.5)
+                    time.sleep(0.5*(attempt**1.2))
                     continue
 
                 if isinstance(e, AnonymousAccessDeniedError):

--- a/datalad/downloaders/base.py
+++ b/datalad/downloaders/base.py
@@ -176,6 +176,12 @@ class BaseDownloader(object, metaclass=ABCMeta):
                 break
             except AccessDeniedError as e:
                 ce = CapturedException(e)
+                if hasattr(e, 'status') and e.status == 429:
+                    # Too many requests.
+                    # We can retry by continuing the loop.
+                    time.sleep(0.5)
+                    continue
+
                 if isinstance(e, AnonymousAccessDeniedError):
                     access_denied = "Anonymous access"
                 else:

--- a/datalad/downloaders/http.py
+++ b/datalad/downloaders/http.py
@@ -121,21 +121,25 @@ def check_response_status(response, err_prefix="", session=None):
         raise AccessDeniedError(
             err_msg,
             supported_types=process_www_authenticate(
-                response.headers.get('WWW-Authenticate')))
+                response.headers.get('WWW-Authenticate')),
+            status=response.status_code)
     elif response.status_code in {200}:
         pass
     elif response.status_code in {301, 302, 307}:
         # TODO: apparently tests do not exercise this one yet
         if session is None:
-            raise AccessFailedError(err_msg + " no session was provided")
+            raise AccessFailedError(err_msg + " no session was provided",
+                                    status=response.status_code)
         redirs = list(session.resolve_redirects(response, response.request))
         if len(redirs) > 1:
             lgr.warning("Multiple redirects aren't supported yet.  Taking first")
         elif len(redirs) == 0:
-            raise AccessFailedError("No redirects were resolved")
-        raise UnhandledRedirectError(err_msg, url=redirs[0].url)
+            raise AccessFailedError("No redirects were resolved",
+                                    status=response.status_code)
+        raise UnhandledRedirectError(err_msg, url=redirs[0].url,
+                                     status=response.status_code)
     else:
-        raise AccessFailedError(err_msg)
+        raise AccessFailedError(err_msg, status=response.status_code)
 
 
 @auto_repr

--- a/datalad/support/exceptions.py
+++ b/datalad/support/exceptions.py
@@ -532,7 +532,11 @@ class ConnectionOpenFailedError(CommandError):
 
 
 class DownloadError(Exception):
-    pass
+
+    def __init__(self, msg=None, status=None, **kwargs):
+        super(DownloadError, self).__init__(msg, **kwargs)
+        # store response status code
+        self.status = status
 
 
 class IncompleteDownloadError(DownloadError):

